pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -1614,68 +1628,111 @@ char **NpyArr_encodeLabels(PyArrayObject *labels, JSONObjectEncoder *enc,
         ret[i] = NULL;
     }
 
-    origst = enc->start;
-    origend = enc->end;
-    origoffset = enc->offset;
-
     stride = PyArray_STRIDE(labels, 0);
     dataptr = PyArray_DATA(labels);
     getitem = (PyArray_GetItemFunc *)PyArray_DESCR(labels)->f->getitem;
     type_num = PyArray_TYPE(labels);
 
     for (i = 0; i < num; i++) {
-        if (PyTypeNum_ISDATETIME(type_num) || PyTypeNum_ISNUMBER(type_num))
-        {
-            item = (PyObject *)labels;
-            pyenc->npyType = type_num;
-            pyenc->npyValue = dataptr;
-        } else {
-            item = getitem(dataptr, labels);
-            if (!item) {
-                NpyArr_freeLabels(ret, num);
-                ret = 0;
-                break;
-            }
-        }
-
-        cLabel = JSON_EncodeObject(item, enc, labelBuffer, NPY_JSON_BUFSIZE);
-
-        if (item != (PyObject *)labels) {
-            Py_DECREF(item);
-        }
+        item = getitem(dataptr, labels);
+        if (!item) {
+	  NpyArr_freeLabels(ret, num);
+	  ret = 0;
+	  break;
+	}
+
+	// Using a date as a key we need to special case the formatting
+	if (PyTypeNum_ISDATETIME(type_num) ||
+	    PyDateTime_Check(item) || PyDate_Check(item)) {
```

Comment:
Might need to add Timedelta to the list though this doesn't fail anything. I'll check more thoroughly next

Reply:
It looks like there is an existing issue with `Timedelta` and how it writes out opened by @TomAugspurger in #15137 

Here is behavior on master:

```python
In [8]: pd.DataFrame([[1]], columns=[pd.Timedelta('1D')]).to_json(date_format="iso")
Out[8]: '{"1970-01-02T00:00:00.000Z":{"0":1}}'
```

versus this branch

```python
In [9]: pd.DataFrame([[1]], columns=[pd.Timedelta('1D')]).to_json(date_format="iso")
Out[9]: '{"1970-01-02T00:00:00":{"0":1}}'
```

Neither is correct though I think this branch would make it easier to identify how to actually make that update, especially once the logic for writing out dt-like objects is aligned between the keys and values of JSON serialization